File: clients/swift/AgentHostProtocol/Sources/AgentHostProtocolClient/Hosts/HostClientHandle.swift
// HostClientHandle — generation-checked escape hatch onto the underlying
// `AHPClient` for a host.

import Foundation
import AgentHostProtocol

/// Generation-checked handle to the underlying single-host `AHPClient`.
///
/// Issued by `MultiHostClient.client(for:)`. Methods on this handle verify
/// that the host is still on the same `generation` it was when the handle
/// was minted; if a reconnect has occurred, dispatching returns
/// `HostError.hostReconnected` instead of silently writing to the new
/// connection.
///
/// **Race note:** generation is checked once at the start of each call, so
/// it is possible (but rare) for a reconnect to land between
/// `checkAlive()` and the actual `dispatch`/`request`. In that race the
/// dispatch goes against the stale (now-shutdown) `AHPClient` and surfaces
/// as `AHPClientError.shutdown` wrapped in `HostError.client`. Acquire a
/// fresh handle when this happens. The Rust SDK has the same semantics.
public struct HostClientHandle: Sendable {
    /// Host this handle was issued for.
    public let hostId: HostId
    /// Generation this handle was minted at.
    public let generation: UInt64

    /// The `AHPClient` instance that was current when the handle was minted.
    /// May have been shut down by a subsequent reconnect.
    private let client: AHPClient
    private let shared: HostShared

    internal init(hostId: HostId, generation: UInt64, client: AHPClient, shared: HostShared) {
        self.hostId = hostId
        self.generation = generation
        self.client = client
        self.shared = shared
    }

    /// Validate this handle against the host's current generation. Throws
    /// `HostError.hostReconnected` if a reconnect has happened.
    public func checkAlive() async throws {
        let current = await shared.generation()
        if current != generation {
            throw HostError.hostReconnected(
                host: hostId,
                handleGeneration: generation,
                currentGeneration: current
            )
        }
    }

    /// Dispatch an action through this connection on `channel`, refusing if
    /// the connection has been replaced by a reconnect.
    @discardableResult
    public func dispatch(_ action: StateAction, channel: String) async throws -> DispatchHandle {
        try await checkAlive()
        do {
            return try await client.dispatch(action, channel: channel)
        } catch let error as AHPClientError {
            throw HostError.client(error)
        }
    }

    /// Dispatch an action on `channel` with a caller-owned `clientSeq`,
    /// refusing if the connection has been replaced by a reconnect.
    @discardableResult
    public func dispatch(_ action: StateAction, channel: String, clientSeq: Int) async throws -> DispatchHandle {
        try await checkAlive()
        do {
            return try await client.dispatch(action, channel: channel, clientSeq: clientSeq)
        } catch let error as AHPClientError {
            throw HostError.client(error)
        }
    }

    /// Issue an arbitrary JSON-RPC request through this connection, refusing
    /// if the connection has been replaced by a reconnect.
    public func request<P: Encodable & Sendable, R: Decodable & Sendable>(
        method: String,
        params: P
    ) async throws -> R {
        try await checkAlive()
        do {
            return try await client.request(method: method, params: params)
        } catch let error as AHPClientError {
            throw HostError.client(error)
        }
    }

    /// Borrow the underlying `AHPClient` for advanced use. The caller is
    /// responsible for not holding it past the next reconnect — the returned
    /// reference becomes a stale handle once the host reconnects.
    public func rawClient() async -> AHPClient {
        client
    }
}

File: clients/dotnet/src/AgentHostProtocol/Hosts/HostClientHandle.cs
// Generation-checked handle onto the underlying single-host AhpClient.
#nullable enable

using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.AgentHostProtocol.Hosts;

/// <summary>
/// Generation-checked handle onto the underlying single-host <see cref="AhpClient"/>
/// for a host. Issued by <see cref="MultiHostClient.ClientFor"/>. Operations verify
/// the host is still registered and on the same <c>Generation</c> the handle was
/// minted at; if the host was removed/shut down they throw
/// <see cref="HostShutDownException"/>, and if a reconnect replaced the connection
/// they throw <see cref="HostNotConnectedException"/> (acquire a fresh handle).
/// Port of Swift's <c>HostClientHandle</c> (Swift surfaces the reconnect case as
/// <c>hostReconnected</c>; the .NET typed-error set folds that into "not the
/// connection you held — reacquire").
/// </summary>
public sealed class HostClientHandle
{
    private readonly MultiHostClient _owner;

    /// <summary>The host this handle was issued for.</summary>
    public HostId HostId { get; }

    /// <summary>The generation this handle was minted at.</summary>
    public ulong Generation { get; }

    internal HostClientHandle(MultiHostClient owner, HostId hostId, ulong generation)
    {
        _owner = owner; HostId = hostId; Generation = generation;
    }

    /// <summary>
    /// Validates this handle and returns the underlying live client. Throws
    /// <see cref="HostShutDownException"/> if the host is no longer registered
    /// (removed or the multi-host client shut down), or
    /// <see cref="HostNotConnectedException"/> if the host has reconnected (the
    /// generation moved) or currently has no live connection.
    /// </summary>
    private AhpClient CheckAlive()
    {
        var entry = _owner.TryGetEntry(HostId);
        if (entry is null) throw new HostShutDownException(HostId);
        var snap = entry.Snapshot();
        if (snap.Generation != Generation) throw new HostNotConnectedException(HostId);
        var client = entry.CurrentClient;
        if (client is null) throw new HostNotConnectedException(HostId);
        return client;
    }

    /// <summary>
    /// Throws if this handle is no longer valid (host removed →
    /// <see cref="HostShutDownException"/>; reconnected/disconnected →
    /// <see cref="HostNotConnectedException"/>). Mirrors Swift's <c>checkAlive()</c>.
    /// </summary>
    public void CheckAliveOrThrow() => CheckAlive();

    /// <summary>
    /// Dispatches an action through this connection on <paramref name="channel"/>,
    /// refusing (throwing) if the host was removed or the connection has been
    /// replaced. Mirrors Swift's <c>HostClientHandle.dispatch</c>.
    /// </summary>
    /// <param name="action">The action to dispatch.</param>
    /// <param name="channel">Channel URI the action targets.</param>
    /// <param name="clientSeq">
    /// Optional caller-owned sequence number. When supplied, that exact value is
    /// sent on the wire and recorded on the returned handle; when <c>null</c>, the
    /// connection's next auto-incrementing sequence is used. Mirrors Swift's
    /// <c>HostClientHandle.dispatch(action:channel:clientSeq:)</c>.
    /// </param>
    /// <param name="cancellationToken">Cancels the send.</param>
    public async Task<DispatchHandle> DispatchAsync(
        StateAction action,
        string channel,
        long? clientSeq = null,
        CancellationToken cancellationToken = default)
    {
        var client = CheckAlive();
        return await client.DispatchAsync(channel, action, clientSeq, cancellationToken).ConfigureAwait(false);
    }
}
